### DISCUSSION TITLE
Scheduling in a WinForms app

### DISCUSSION THREAD

**Discussion by simmotech** [quartznet/quartznet#1301] [General]
I've been browsing source code most of the day looking to replace the existing scheduling in my WinForms app with something more comprehensive and robust.
Quartz.Net seems to cover a lot of Scheduling scenarios (90% of them I won't need but its nice to know they are there) but built-in support for simpler scheduling scenarios is not quite there.

As an example, my WinForms app has Widgets - basically just user controls that show some information and have a Refresh button for manual updating but also allow periodical updating in the background. Because these already exist in the GUI, the notion of creating an IJob instance is a non-starter for this. I basically just want to use a Trigger to run the instance method that refreshes the data.

I have started playing with a custom JobFactory that overrides NewJob() and uses this code
`bundle.JobDetail.JobDataMap["Instance"] as IJob ?? base.NewJob(bundle, scheduler);
to use an existing instance. This also gets around the parameterless ctor restriction.

Then there is the question of running the widget refresh on a background thread or the GUI thread.
Currently I use a background thread and use BeginInvoke before and after the actual fetch to update the GUI but I am also now trying to use Async since a lot of my Widgets would be able to refetch their data that way.

So I'm thinking that my App needs two Scheduler instances - one that uses StdSchedulerFactory for the background thread stuff and another that will call IJob.Execute() on the GUI thread.

I think that I could derive a custom TaskSchedulingThreadPool like this

```
        protected override TaskScheduler GetDefaultScheduler()
        {
            return TaskScheduler.FromCurrentSynchronizationContext();
        }

```

Am I on the right lines with two Schedulers?
Anything else I need to watch out for?

Thanks.

**Comment by lahma** [comment]
Unfortunately I don't have much guidance here. I personally haven't used Quartz.NET with WinForms, but I do know that some have. You are probably on the right track with custom thread pool, but I would also consider whether Quartz.NET is too "heavy-weight" compared to using something simpler, like `Timer`s. Especially if the case is just to poll things on fixed intervals.

> So I'm thinking that my App needs two Scheduler instances - one that uses StdSchedulerFactory for the background thread stuff and another that will call IJob.Execute() on the GUI thread.

Another option might be to separate the background stull for Quartz.NET (it's really good there!) and leave UI updates to timer or similar construct?
